src/Database/LSMTree/Internal/Unsliced.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE DeriveFunctor     #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RoleAnnotations   #-}
+
+module Database.LSMTree.Internal.Unsliced (
+    -- * Unsliced raw bytes
+    Unsliced
+    -- * Unsliced keys
+  , makeUnslicedKey
+  , unsafeMakeUnslicedKey
+  , fromUnslicedKey
+  , unsafeNoAssertMakeUnslicedKey
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Primitive.ByteArray
+import qualified Data.Vector.Primitive as PV
+import           Database.LSMTree.Internal.RawBytes (RawBytes (..))
+import qualified Database.LSMTree.Internal.RawBytes as RB
+import           Database.LSMTree.Internal.Serialise (SerialisedKey (..))
+import           Database.LSMTree.Internal.Vector (mkPrimVector,
+                     noRetainedExtraMemory)
+
+-- | Unsliced string of bytes
+type role Unsliced nominal
+newtype Unsliced a =
+    Unsliced ByteArray
+
+getByteArray :: RawBytes -> ByteArray
+getByteArray (RawBytes (PV.Vector _ _ ba)) = ba
+
+precondition :: RawBytes -> Bool
+precondition (RawBytes pvec) = noRetainedExtraMemory pvec
+
+makeUnsliced :: RawBytes -> Unsliced RawBytes
+makeUnsliced bytes
+    | precondition bytes = Unsliced (getByteArray bytes)
+    | otherwise     = Unsliced (getByteArray $ RB.force bytes)
+
+unsafeMakeUnsliced :: RawBytes -> Unsliced RawBytes
+unsafeMakeUnsliced bytes = assert (precondition bytes) (Unsliced (getByteArray bytes))
+
+unsafeNoAssertMakeUnsliced :: RawBytes -> Unsliced RawBytes
+unsafeNoAssertMakeUnsliced bytes = Unsliced (getByteArray bytes)
+
+fromUnsliced :: Unsliced RawBytes -> RawBytes
+fromUnsliced (Unsliced ba) = RawBytes (mkPrimVector 0 (sizeofByteArray ba) ba)
+
+{-------------------------------------------------------------------------------
+  Unsliced keys
+-------------------------------------------------------------------------------}
+
+from :: Unsliced RawBytes -> Unsliced SerialisedKey
+from (Unsliced ba) = Unsliced ba
+
+to :: Unsliced SerialisedKey -> Unsliced RawBytes
+to (Unsliced ba) = Unsliced ba
+
+makeUnslicedKey :: SerialisedKey -> Unsliced SerialisedKey
+makeUnslicedKey (SerialisedKey rb) = from (makeUnsliced rb)
+
+unsafeMakeUnslicedKey :: SerialisedKey -> Unsliced SerialisedKey
+unsafeMakeUnslicedKey (SerialisedKey rb) = from (unsafeMakeUnsliced rb)
+
+unsafeNoAssertMakeUnslicedKey :: SerialisedKey -> Unsliced SerialisedKey
+unsafeNoAssertMakeUnslicedKey (SerialisedKey rb) = from (unsafeNoAssertMakeUnsliced rb)
+
+fromUnslicedKey :: Unsliced SerialisedKey -> SerialisedKey
+fromUnslicedKey x = SerialisedKey (fromUnsliced (to x))
+
+instance Show (Unsliced SerialisedKey) where
+  show x = show (fromUnslicedKey x)
+
+instance Eq (Unsliced SerialisedKey) where
+  x == y = fromUnslicedKey x == fromUnslicedKey y
+
+instance Ord (Unsliced SerialisedKey) where
+  x <= y = fromUnslicedKey x <= fromUnslicedKey y
```

Comment:
Why are these on `Unsliced SerialisedKey` specifically rather than derived Eq, Ord, Show for all types?

Reply:
I suppose I could add a function like that akin to

`foo :: Coercible RawBytes a => Unsliced Rawbytes -> a`

Reply:
I'm going to merge it as is, if we want to add more instances or make these instances more general, we can do that once the need arises